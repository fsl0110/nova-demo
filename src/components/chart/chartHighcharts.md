### A Highcharts-Component

#### Chart without error

```jsx
const testProps = {
  location: "initiators",
  fetchHandler: () => null,
  chartOptions: {
    chart: {
      height: 350,
      width: 350,
      style: {
        fontFamily:
          "-apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen"
      }
    },
    credits: {
      enabled: false
    },
    loading: {
      hideDuration: 500,
      showDuration: 300
    },
    title: {
      text: ""
    },
    series: [
      {
        type: "pie",
        data: [
          ["Apples", 29.9, false],
          ["Pears", 71.5, false],
          ["Oranges", 106.4, false],
          ["Plums", 129.2, false],
          ["Bananas", 144.0, false],
          ["Peaches", 176.0, false],
          ["Prunes", 135.6, true, true],
          ["Avocados", 148.5, false]
        ]
      }
    ]
  }
};

<ChartHighcharts {...testProps} />;
```

#### Chart with error

```jsx
const testProps = {
  location: "initiators",
  hasError: true,
  fetchHandler: () => null,
  chartOptions: {}
};

<ChartHighcharts {...testProps} />;
```

#### Chart while loading

```jsx
const testProps = {
  location: "initiators",
  isLoading: true,
  fetchHandler: () => null,
  chartOptions: {}
};

<ChartHighcharts {...testProps} />;
```
