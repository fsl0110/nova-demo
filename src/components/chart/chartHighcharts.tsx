import React, { HTMLAttributes, PureComponent } from "react";
import Highcharts from "highcharts";
import HighchartsReact from "highcharts-react-official";
import classNames from "classnames";
import { Error } from "..";
require("highcharts-no-data-to-display")(Highcharts);

export interface Props extends HTMLAttributes<HTMLDivElement> {
  location: string;
  /** Is loading data. */
  isLoading?: boolean;
  /** Has an error. */
  hasError?: boolean;
  /** Options and Data of the chart */
  chartOptions: Highcharts.Options;
  /** Action for fetching data. */
  fetchHandler: (location: string) => void;
}

export class ChartHighcharts extends PureComponent<Props> {
  private chartInstance: any;

  constructor(props: Props) {
    super(props);

    this.afterChartCreated = this.afterChartCreated.bind(this);
  }

  afterChartCreated(chart: any) {
    this.chartInstance = chart;
  }

  componentDidMount() {
    this.props.fetchHandler(this.props.location);
  }

  render() {
    const {
      location,
      isLoading,
      hasError,
      chartOptions,
      fetchHandler,
      className
    } = this.props;

    if (!hasError && this.chartInstance && isLoading) {
      this.chartInstance.showLoading();
    }
    if (!hasError && this.chartInstance && !isLoading) {
      this.chartInstance.hideLoading();
    }

    return (
      <div className={classNames(className)}>
        {!hasError ? (
          <HighchartsReact
            highcharts={Highcharts}
            options={chartOptions}
            callback={this.afterChartCreated}
          />
        ) : (
          <Error
            location={location}
            fetchData={fetchHandler}
            isLoading={isLoading}
          />
        )}
      </div>
    );
  }
}

export default ChartHighcharts;
