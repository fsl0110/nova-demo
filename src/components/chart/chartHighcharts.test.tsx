import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { ChartHighcharts, Props } from "./chartHighcharts";

const testProps: Props = {
  location: "reports",
  fetchHandler: jest.fn(),
  chartOptions: {
    chart: {
      height: 350,
      width: 350,
      style: {
        fontFamily:
          "-apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen"
      }
    },
    credits: {
      enabled: false
    },
    loading: {
      hideDuration: 500,
      showDuration: 300
    },
    title: {
      text: ""
    },
    series: [{ type: "pie", data: [null] }]
  }
};

describe("<ChartHighcharts>", () => {
  it("renders without crashing", () => {
    shallow(<ChartHighcharts {...testProps} />);
  });

  it("renders with error", () => {
    const newTestProps = { ...testProps, hasError: true };
    const mock = shallow(<ChartHighcharts {...newTestProps} />);
    expect(toJson(mock)).toMatchSnapshot();
  });
});
