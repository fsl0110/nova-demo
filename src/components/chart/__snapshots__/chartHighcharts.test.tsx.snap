// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChartHighcharts> renders with error 1`] = `
<div
  className=""
  data-test="ChartHighcharts"
>
  <Error
    fetchData={
      [MockFunction] {
        "calls": Array [
          Array [
            "reports",
          ],
          Array [
            "reports",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    location="reports"
  />
</div>
`;
